tests/mocks/core/services/request/RequestMock.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+namespace EventEspresso\tests\mocks\core\services\request;
+
+use EventEspresso\core\domain\services\contexts\RequestTypeContextCheckerInterface;
+use EventEspresso\core\services\request\Request;
+
+/**
+ * Class RequestMock
+ * Description
+ *
+ * @package EventEspresso\tests\mocks\core\services\request
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+class RequestMock extends Request
+{
+
+    /**
+     * $_GET parameters
+     *
+     * @var array $get
+     */
+    public $get;
+
+    /**
+     * $_POST parameters
+     *
+     * @var array $post
+     */
+    public $post;
+
+    /**
+     * $_COOKIE parameters
+     *
+     * @var array $cookie
+     */
+    public $cookie;
+
+    /**
+     * $_SERVER parameters
+     *
+     * @var array $server
+     */
+    public $server;
+
+    /**
+     * $_REQUEST parameters
+     *
+     * @var array $request
+     */
+    public $request;
+
+    /**
+     * @var RequestTypeContextCheckerInterface
+     */
+    public $request_type;
+
+    /**
+     * IP address for request
+     *
+     * @var string $ip_address
+     */
+    public $ip_address;
+
+    /**
+     * @var string $user_agent
+     */
+    public $user_agent;
+
+    /**
+     * true if current user appears to be some kind of bot
+     *
+     * @var bool $is_bot
+     */
+    public $is_bot;
+
+
+    public function __construct(array $get, array $post, array $cookie, array $server, $ip_address = '0.0.0.0')
+    {
+        $this->get = $get;
+        $this->post = $post;
+        $this->cookie = $cookie;
+        $this->server = $server;
+        $this->request = array_merge($this->get, $this->post);
+        $this->ip_address = $this->visitorIp($ip_address);
+        parent::__construct($get, $post, $cookie, $server);
+    }
+
+
+    /**
+     * @param string $key
+     * @param string $value
+     * @param bool   $override_ee
+     * @return    void
+     */
+    public function setRequestParam($key, $value, $override_ee = false)
+    {
+        $this->request[ $key ] = $value;
+    }
+
+
+    /**
+     * returns the value for a request param if the given key exists
+     *
+     * @param string $key
+     * @param null   $default
+     * @return mixed
+     */
+    public function getRequestParam($key, $default = null)
+    {
+        return isset($this->request[ $key ]) ? $this->request[ $key ] : $default;
```

Comment:
Does overriding the original method mean that the private method `RequestParameterDrillDown` is not covered by tests?

- This applies also to `requestParamIsSet`.
- This applies also to `getMatch`.

Reply:
The purpose of the `RequestMock` class is NOT for testing the `Request` class. Its purpose is so that you can remove the effects of the `Request` class from other tests. In this case we are removing the effects of the `Request` class from the tests in the `RequestTypeContextDetectorTest` test case.

The private `RequestParameterDrillDown` method is (indirectly) covered in the tests in `EventEspresso\tests\testcases\core\request_stack\RequestTest`, specifically by the following tests:

- `testGetWithDrillDown()`
- `testIsSetWithDrillDown()`
- `testMatchesWithDrillDown()`
- `testGetMatchWithDrillDown()`